activerecord/test/support/method_stub.rb

```diff
@@ -0,0 +1,54 @@
+# frozen_string_literal: true
+
+module MethodStub
+  # FIXME: Replace tests for method calls (and not called), especially
+  # where the method stubbed is not defined for the object, with simpler
+  # tests.
+  private
+
+    def with_stubbed_method(object, method_name, return_value = nil)
+      stub_method(object, method_name, return_value)
+
+      yield
+
+    ensure
+      unstub_method(object, method_name)
+    end
+
+    # TODO: `alias_method` is private in Ruby 2.4, so once the minimum
+    # version is 2.5 there'll be no need to resort to `module_eval`.
```

Comment:
Could you elaborate how it will prevent using of `module_eval` here?
Using `module_eval` seems fine to me here.

Reply:
In `RUBY_VERSION > "2.5"` `alias_method` is public and we can avoid resorting to `module_eval` like so:

```
    def stub_method(object, method_name, return_value = nil)
      base_var_name = method_name.to_s.gsub(/[?!]/, "")

      object.class_variable_set("@@#{base_var_name}_called", 0)

      object.define_method("stubbed_#{method_name}") do |*args|
        object.class_variable_set(
          "@@#{base_var_name}_called",
          object.class_variable_get("@@#{base_var_name}_called") + 1
        )

        return_value
      end

      if object.method_defined?(method_name)
        object.alias_method("original_#{method_name}", method_name)
      end
      object.alias_method(method_name, "stubbed_#{method_name}")
    end

    def unstub_method(object, method_name)
      base_var_name = method_name.to_s.gsub(/[?!]/, "")

      if object.method_defined?("original_#{method_name}")
        object.alias_method(method_name, "original_#{method_name}")
        object.remove_method("original_#{method_name}")
      else
        object.remove_method(method_name)
      end
      object.remove_method("stubbed_#{method_name}")
      object.remove_class_variable("@@#{base_var_name}_called")
    end
```

Reply:
It technically it would be the same, so I'm not sure whether it worth to add this `TODO`
@utilum I've updated my diff (https://github.com/rails/rails/pull/33162#issuecomment-407213833) by adding more tests for these methods and improved the methods.

Reply:
It would affect the same modifications to `object`, but it's easier to debug. Also it would be evaluated and checked in `ruby -cw`, as well as validated by `rubocop`.